workflow/controller/dag.go

```diff
@@ -526,7 +526,12 @@ func (woc *wfOperationCtx) resolveDependencyReferences(dagCtx *dagContext, task
 		return nil, errors.InternalWrapError(err)
 	}
 	fstTmpl := fasttemplate.New(string(taskBytes), "{{", "}}")
-	newTaskStr, err := common.Replace(fstTmpl, scope.replaceMap(), true)
+
+	replaceMap := scope.replaceMap()
```

Comment:
I think something like 
```suggestion
	replaceMap := scope.replaceMap(woc.globalParams)
```

Would be cleaner.

You could edit the `replaceMap` function to include this optional parameter and add its contents to the scope if it's not nil

Reply:
What do you think?

Reply:
I think the method name `replaceMap` is not intuitive to accept an argument of params extension. Maybe, we can have something like `mergeParams` utility method in the common package.

Reply:
If we want to introduce `mergeParams` utility method, we should make another PR to replace other codes meging params.

Reply:
We could change the name to something like `prepareScope`, here is an example where I do something similar:

https://github.com/argoproj/argo/blob/7cb2fd17765aad691eda25ca4c5acecb89f84394/workflow/controller/steps.go#L466-L471

This is minor though, I'll leave it up to you.

Reply:
As I found there're many cases of merging params, I defined `Params` type which has `Clone` and `Merge` functions so we can replace those codes with one line.

I also found the order of merging was not correct in my original commit of this PR. `scoep.replaceMap()` should be merged into `globalParmas` but not the opposite.